src/sidebar/services/auth.ts

```diff
@@ -0,0 +1,331 @@
+    toastMessenger: ToastMessengerService
+  ) {
+    super();
+
+    this._authCode = null;
+    this._client = null;
+    this._tokenInfoPromise = null;
+
+    this._apiRoutes = apiRoutes;
+    this._localStorage = localStorage;
+    this._settings = settings;
+    this._toastMessenger = toastMessenger;
+    this._window = $window;
+
+    this._listenForTokenStorageEvents();
+  }
+
+  /**
+   * Listen for updated access and refresh tokens saved by other instances of
+   * the client.
+   */
+  private _listenForTokenStorageEvents() {
+    this._window.addEventListener('storage', ({ key }) => {
+      if (key === this._storageKey()) {
+        // Reset cached token information. Tokens will be reloaded from storage
+        // on the next call to `getAccessToken()`.
+        this._tokenInfoPromise = null;
+        this.emit('oauthTokensChanged');
+      }
+    });
+  }
+
+  /**
+   * Fetch the last-saved access/refresh tokens from local storage.
+   */
+  private _loadToken() {
+    const token = this._localStorage.getObject(this._storageKey());
+
+    if (
+      !token ||
+      typeof token.accessToken !== 'string' ||
+      typeof token.refreshToken !== 'string' ||
+      typeof token.expiresAt !== 'number'
+    ) {
+      return null;
+    }
+
+    return {
+      accessToken: token.accessToken,
+      refreshToken: token.refreshToken,
+      expiresAt: token.expiresAt,
+    };
+  }
+
+  /**
+   * Return a cached OAuthClient if available, or configure one and return it.
+   */
+  async _oauthClient() {
+    if (this._client) {
+      return this._client;
+    }
+
+    const links = await this._apiRoutes.links();
+    this._client = new OAuthClient({
+      clientId: this._settings.oauthClientId,
+      authorizationEndpoint: links['oauth.authorize'],
+      revokeEndpoint: links['oauth.revoke'],
+      tokenEndpoint: resolve('token', this._settings.apiUrl),
+    });
+    return this._client;
+  }
+
+  /**
+   * Exchange the authorization code retrieved from login popup for a new
+   * access token.
+   */
+  private async _exchangeAuthCodeForToken(code: string) {
+    const client = await this._oauthClient();
+    const tokenInfo = await client.exchangeAuthCode(code);
+    this._saveToken(tokenInfo);
+    return tokenInfo;
+  }
+
+  /**
+   * Attempt to exchange a grant token for an access token.
+   *
+   * Publishers or other embedders can set a grant token for a user in service
+   * configuration.
+   */
+  private async _exchangeGrantToken(grantToken: string) {
+    const client = await this._oauthClient();
+    try {
+      return client.exchangeGrantToken(grantToken);
+    } catch (err) {
+      this._toastMessenger.error(
+        `Hypothesis login lost: You must reload the page to annotate.`,
+        {
+          autoDismiss: false,
+        }
+      );
+      throw err;
+    }
+  }
+
+  /**
+   * Retrieve any grant token set by service configuration
+   */
+  private _getGrantToken(): string | undefined {
```

Comment:
To check my understanding, there are three possibilities here:

1. `undefined` - We're not using the grant token login method
2. Empty string - We are using the grant token login method, but the user is logged out
3. Non-empty string - We are using the grant token login method, and we should attempt a login

Is this correct?

Reply:
Correct.

Reply:
> To check my understanding, there are three possibilities here:
> 
> 1. `undefined` - We're not using the grant token login method
> 2. Empty string - We are using the grant token login method, but the user is logged out
> 3. Non-empty string - We are using the grant token login method, and we should attempt a login
> 
> Is this correct?

Yes, clarifying that "we should attempt a login" means "we should attempt to exchange the grant token for an access token"